src/write_redis.c

```diff
@@ -26,182 +26,340 @@
+    return ENOTCONN;
+  }
+  if (node->conn->err) {
+    ERROR("write_redis plugin: Connecting to host \"%s\" (port %i) failed: %s",
+          (node->host != NULL) ? node->host : "localhost",
+          (node->port != 0) ? node->port : REDIS_DEFAULT_PORT,
+          node->conn->errstr);
+    disconnect(node);
+    return ENOTCONN;
+  }
 
-    rr = redisCommand(node->conn, "SELECT %d", node->database);
-    if (rr == NULL)
-      WARNING("SELECT command error. database:%d message:%s", node->database,
-              node->conn->errstr);
-    else
-      freeReplyObject(rr);
+  redisReply *rr = redisCommand(node->conn, "SELECT %d", node->database);
+  if (rr == NULL) {
+    ERROR("write_redis plugin: Command \"SELECT %d\" failed: %s",
+          node->database, node->conn->errstr);
+    disconnect(node);
+    return ENOTCONN;
   }
 
-  rr = redisCommand(node->conn, "ZADD %s %s %s", key, time, value);
-  if (rr == NULL)
-    WARNING("ZADD command error. key:%s message:%s", key, node->conn->errstr);
-  else
+  freeReplyObject(rr);
+  return 0;
+}
+
+static void print_execute_error(int argc, char const **argv, char const *msg) {
+  strbuf_t cmd = STRBUF_CREATE;
+  for (int i = 0; i < argc; i++) {
+    if (i != 0) {
+      strbuf_print(&cmd, " ");
+    }
+    strbuf_print(&cmd, argv[i]);
+  }
+
+  ERROR("write_redis plugin: Command \"%s\" failed: %s", cmd.ptr, msg);
+
+  STRBUF_DESTROY(cmd);
+}
+
+static int execute(wr_node_t *node, int argc, char const **argv) {
+  redisReply *rr = redisCommandArgv(node->conn, argc, argv, NULL);
+  if (rr == NULL) {
+    print_execute_error(argc, argv, node->conn->errstr);
+    return (node->conn->err != 0) ? node->conn->err : -1;
+  }
+  if (rr->type == REDIS_REPLY_ERROR) {
+    print_execute_error(argc, argv, rr->str);
     freeReplyObject(rr);
+    return -1;
+  }
 
-  if (node->max_set_size >= 0) {
-    rr = redisCommand(node->conn, "ZREMRANGEBYRANK %s %d %d", key, 0,
-                      (-1 * node->max_set_size) - 1);
-    if (rr == NULL)
-      WARNING("ZREMRANGEBYRANK command error. key:%s message:%s", key,
-              node->conn->errstr);
-    else
-      freeReplyObject(rr);
+  freeReplyObject(rr);
+  return 0;
+}
+
+static int add_resource_to_global_set(wr_node_t *node, char const *id) {
+  strbuf_t key = STRBUF_CREATE;
+  if (node->prefix != NULL) {
+    strbuf_print(&key, node->prefix);
   }
+  strbuf_print(&key, "resources");
 
-  if (node->max_set_duration > 0) {
-    /*
-     * remove element, scored less than 'current-max_set_duration'
-     * '(...' indicates 'less than' in redis CLI.
-     */
-    rr = redisCommand(node->conn, "ZREMRANGEBYSCORE %s -1 (%.9f", key,
-                      (CDTIME_T_TO_DOUBLE(m->time) - node->max_set_duration));
-    if (rr == NULL)
-      WARNING("ZREMRANGEBYSCORE command error. key:%s message:%s", key,
-              node->conn->errstr);
-    else
-      freeReplyObject(rr);
+  char const *cmd[] = {"SADD", key.ptr, id};
```

Comment:
`ZADD` -> `SADD`?

(I have no idea about Redis commands, but noticed that command name had changed.)

Reply:
`SADD` is correct here. `ZADD` is used for a sorted set, which requires a "score". This set is not sorted.

The diff is displaying a bit weird. This is taking the place of the `SADD` command in line 153 (adding `metric_string_p`).